src/helpers/network.rs

```diff
@@ -1,101 +1,80 @@
+pub struct Network<T> {
+    transport: T,
+    query_id: QueryId,
+    roles_to_helpers: HashMap<Role, HelperIdentity>,
 }
 
-/// Wrapper around a [`PollSender`] to modify the error message to match what the [`NetworkSink`]
-/// requires. The only error that [`PollSender`] will generate is "channel closed", and thus is the
-/// only error message forwarded from this [`NetworkSink`].
-#[pin_project]
-pub struct NetworkSink<T> {
-    #[pin]
-    inner: PollSender<T>,
-}
-
-impl<T: Send + 'static> NetworkSink<T> {
-    #[must_use]
-    pub fn new(sender: mpsc::Sender<T>) -> Self {
+impl<T: Transport> Network<T> {
+    pub fn new(
+        transport: T,
+        query_id: QueryId,
+        roles_to_helpers: HashMap<Role, HelperIdentity>,
+    ) -> Self {
         Self {
-            inner: PollSender::new(sender),
+            transport,
+            query_id,
+            roles_to_helpers,
         }
     }
-}
 
-impl<T: Send + 'static> futures::Sink<T> for NetworkSink<T>
-where
-    Error: From<PollSendError<T>>,
-{
-    type Error = Error;
-
-    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        ready!(self.project().inner.poll_ready(cx)?);
-        Poll::Ready(Ok(()))
-    }
-
-    fn start_send(self: Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
-        self.project().inner.start_send(item)?;
-        Ok(())
+    /// sends a [`NetworkEvent`] containing [`MessageChunks`] on the underlying [`Transport`]
+    pub async fn send(&self, message_chunks: MessageChunks) -> Result<(), Error> {
+        let role = message_chunks.0.role;
+        let destination = self.roles_to_helpers.get(&role).unwrap();
+        self.transport
+            .send(
+                destination,
+                TransportCommand::NetworkEvent(NetworkEventData {
+                    query_id: self.query_id,
+                    roles_to_helpers: self.roles_to_helpers.clone(),
+                    message_chunks,
+                }),
+            )
+            .await
+            .map_err(Error::from)
     }
 
-    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        ready!(self.project().inner.poll_flush(cx)?);
-        Poll::Ready(Ok(()))
-    }
+    /// returns a [`Stream`] of [`MessageChunks`]s from the underlying [`Transport`]
+    /// # Panics
+    /// if called more than once during the execution of a query.
+    pub fn recv_stream(&self) -> impl Stream<Item = MessageChunks> {
+        let query_id = self.query_id;
+        let query_command_stream = self.transport.subscribe(SubscriptionType::Query(query_id));
 
-    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        ready!(self.project().inner.poll_close(cx))?;
-        Poll::Ready(Ok(()))
+        #[allow(unreachable_patterns)] // there will be more commands in the future
+        query_command_stream.map(move |command| match command {
```

Comment:
nice, there is no penalty of having `Network` as an extra layer here (I believe `map` is zero-cost, could be wrong though)